core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -428,7 +428,7 @@ class IncrementalFetchContext(private val time: Time,
         val mustRespond = cachedPart.maybeUpdateResponseData(respData, updateFetchContextAndRemoveUnselected)
         if (mustRespond) {
           nextElement = element
-          if (updateFetchContextAndRemoveUnselected) {
+          if (updateFetchContextAndRemoveUnselected && FetchResponse.recordsSize(respData) > 0) {
```

Comment:
Does it result in the side effect that we always fetch the no-data partitions?

Reply:
@chia7712 I believe it only affects the order of the fetching on the following fetch. Previously those partitions would be penalized in fetch order if their hwm or log start offset updated and there was no data to be returned.

All of the partitions where the fetch metadata had been updated without any data to be returned will end up behind the partitions ordered prior to them that were part of the full fetch response.

Reply:
Thanks for nice explanation!